juju/juju
# raft: Recover cluster when leader of a 2-machine controller is removed

## Description of change

The usual situation for a Juju controller running in high-availability configuration is to have 3 machines (or maybe 5). In some cases though, a controller can be running with 2 machines in the cluster - for example if the operator is trying to go back to one machine, or one machine of a 3-machine controller has failed and a new one hasn't been added yet. This causes problems with Raft: if you have a cluster with 2 voting nodes and one fails, new log entries can't be added. In this situation you can't change the configuration through the Raft commands - there's no way for the log entry to be accepted by a quorum of nodes, since (2/2) + 1 == 2.

We reduce the risk of this configuration by demoting the follower machine - if the follower fails, the cluster can still run (see #8699). This change handles recovery in the case when the leader of a 2-machine cluster fails (and is subsequently removed).

It adds a new worker, the raft-backstop. This sees the update from the peer-grouper indicating that it is the only remaining API server machine, and detects that the remaining machine is non-voting or can't get quorum in the Raft configuration. It force-appends a new configuration entry to the raft log setting this machine as the only voter. The raft worker then times out waiting for contact from a leader, restarts, sees the new configuration and becomes the leader so processing can continue.

This change implements the behaviour in the [Raft Clustering Scenarios](https://docs.google.com/document/d/1shKBRWw-9uvXVfIoAuQhj_19LdbfP2KTmgDWfw_lmzk/edit) doc. 

## QA steps

Bootstrap a controller and enable HA. Use the juju-engine-report command on the machines to determine the raft leader. Remove that machine. See that the remaining machines elect a new leader and the follower is marked as non-voting. Remove the leader again. See that the backstop worker recovers the cluster allowing the remaining machine to become the leader.
